app/models/api_key.rb

```diff
@@ -3,20 +3,68 @@
 class ApiKey < ApplicationRecord
   include PaperTrailTraceable
 
+  RESOURCES = %w[
+    add_on analytic billable_metric coupon applied_coupon credit_note customer_usage
+    customer fee invoice organization payment_request plan subscription lifetime_usage
+    tax wallet wallet_transaction webhook_endpoint
+  ].freeze
+
+  MODES = %w[read write].freeze
+
+  attribute :permissions, default: -> { default_permissions }
+
   belongs_to :organization
 
   before_create :set_value
 
   validates :value, uniqueness: true
   validates :value, presence: true, on: :update
+  validates :permissions, presence: true
+  validate :permissions_keys_compliance
+  validate :permissions_values_allowed
 
   default_scope { active }
 
   scope :active, -> { where('expires_at IS NULL OR expires_at > ?', Time.current) }
   scope :non_expiring, -> { where(expires_at: nil) }
 
+  def permit?(resource, mode)
+    return true unless organization.premium_integrations.include?('api_permissions')
+
+    permissions[resource].include?(mode)
+  end
+
+  def self.default_permissions
+    RESOURCES.index_with { MODES.dup }
+  end
+
   private
 
+  def permissions_keys_compliance
```

Comment:
This means that adding a resource would invalidate all api keys right?

Reply:
Didn't think about this way.

I'd probably expect that if we add new resource we need think about permissions and update all API keys to reflect new changes. If we didn't update API key permissions that would indeed make all of them invalid which shouldn't prevent from using the key but probably will block update/rotate actions until new resource added to the key.

We can relax validation to allow resource be missing in a list which make it less scary adding new resources later. `#permit?` then need to be adjusted so that missing resource always return no access.

Reply:
that sounds like a good idea 👍